File: IOS/Model/BroadcastManager.m
// Created by Michael Simms on 4/8/14.
// Copyright (c) 2014 Michael J. Simms. All rights reserved.

// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

#import "BroadcastManager.h"
#import "Accelerometer.h"
#import "ActivityMgr.h"
#import "ActivityAttribute.h"
#import "LocationSensor.h"
#import "Notifications.h"
#import "Preferences.h"
#import "Urls.h"

#if !OMIT_BROADCAST

#define MESSAGE_ERROR_SENDING NSLocalizedString(@"Error sending to the server", nil)

@implementation BroadcastManager

- (id)init
{
	if (self = [super init])
	{
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(accelerometerUpdated:) name:@NOTIFICATION_NAME_ACCELEROMETER object:nil];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(locationUpdated:) name:@NOTIFICATION_NAME_LOCATION object:nil];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(activityStarted:) name:@NOTIFICATION_NAME_ACTIVITY_STARTED object:nil];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(activityStopped:) name:@NOTIFICATION_NAME_ACTIVITY_STOPPED object:nil];

		self->locationCache = [[NSMutableArray alloc] init];
		self->accelerometerCache = [[NSMutableArray alloc] init];
		self->lastCacheFlush = 0;
		self->errorSending = FALSE;
		self->deviceId = NULL;
	}
	return self;
}

- (void)dealloc
{
	[self flushGlobalBroadcastCacheRest:FALSE withActivityId:[[NSString alloc] initWithFormat:@"%s", GetCurrentActivityId()]];
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)setDeviceId:(NSString*)deviceId
{
	self->deviceId = deviceId;
}

- (void)displayMessage:(NSString*)text
{	
	NSDictionary* msgData = [[NSDictionary alloc] initWithObjectsAndKeys:text, @KEY_NAME_MESSAGE, nil];

	[[NSNotificationCenter defaultCenter] postNotificationName:@NOTIFICATION_NAME_PRINT_MESSAGE object:msgData];
}

- (void)updateBroadcastStatus:(BOOL)status
{
	NSNumber* numStatus = [[NSNumber alloc] initWithBool:status];
	NSDictionary* msgData = [[NSDictionary alloc] initWithObjectsAndKeys:numStatus, @KEY_NAME_STATUS, nil];

	[[NSNotificationCenter defaultCenter] postNotificationName:@NOTIFICATION_NAME_BROADCAST_STATUS object:msgData];
}

- (void)sendToServer:(NSString*)hostName withPath:(const char*)path withData:(NSMutableData*)data withActivityId:(NSString*)activityId withActivityState:(BOOL)activityStopped
{
	NSString* protocolStr = [Preferences broadcastProtocol];
	NSString* urlStr = [NSString stringWithFormat:@"%@://%@/%s", protocolStr, hostName, path];
	NSString* postLength = [NSString stringWithFormat:@"%lu", (unsigned long)[data length]];

	NSMutableURLRequest* request = [[NSMutableURLRequest alloc] init];
	request.timeoutInterval = 30.0;
	request.allowsExpensiveNetworkAccess = TRUE;
	[request setURL:[NSURL URLWithString:urlStr]];
	[request setHTTPMethod:@"POST"];
	[request setValue:postLength forHTTPHeaderField:@"Content-Length"];
	[request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
	[request setHTTPBody:data];

	self->dataBeingSent = data;

	NSURLSession* session = [NSURLSession sharedSession];
	NSURLSessionDataTask* dataTask = [session dataTaskWithRequest:request
												completionHandler:^(NSData* data, NSURLResponse* response, NSError* error)
	{
		NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*)response;

		if ([httpResponse statusCode] == 200)
		{
			[self updateBroadcastStatus:TRUE];
			self->dataBeingSent = nil;
			self->errorSending = FALSE;

			if (activityStopped)
			{
				dispatch_async(dispatch_get_main_queue(),^{
					[[NSNotificationCenter defaultCenter] postNotificationName:@NOTIFICATION_NAME_BROADCAST_MGR_SENT_ACTIVITY object:activityId];
				} );
			}
		}
		else
		{
			[self displayMessage:MESSAGE_ERROR_SENDING];
			[self updateBroadcastStatus:FALSE];
			self->errorSending = TRUE;
		}
	}];
	[dataTask resume];
}

- (void)flushGlobalBroadcastCacheRest:(BOOL)activityStopped withActivityId:(NSString*)activityId
{
	// No host name set, just return.
	NSString* hostName = [Preferences broadcastHostName];
	if (hostName == nil)
	{
		NSLog(@"Broadcast host name not specified.");
		return;
	}

	// Still waiting on last data to be sent.
	if (self->dataBeingSent)
	{
		if (self->errorSending)
		{
			[self sendToServer:hostName withPath:REMOTE_API_UPDATE_STATUS_URL withData:self->dataBeingSent withActivityId:activityId withActivityState:activityStopped];
			NSLog(@"Resending.");
		}
		else
		{
			NSLog(@"Waiting on previous data to be sent.");
		}
		self->lastCacheFlush = time(NULL);
		return;
	}

	// Write cached location data to the JSON string.
	NSString* post = [NSString stringWithFormat:@"{\"locations\": ["];
	if (!post)
	{
		NSLog(@"Out of memory.");
		self->lastCacheFlush = time(NULL);
		return;
	}
	NSMutableData* postData = [[post dataUsingEncoding:NSASCIIStringEncoding allowLossyConversion:YES] mutableCopy];
	if (!postData)
	{
		NSLog(@"Out of memory.");
		self->lastCacheFlush = time(NULL);
		return;
	}
	size_t numLocObjsBeingSent = 0;
	for (NSString* text in self->locationCache)
	{
		if (numLocObjsBeingSent > 0)
			[postData appendData:[[NSString stringWithFormat:@",\n"] dataUsingEncoding:NSASCIIStringEncoding]];
		[postData appendData:[text dataUsingEncoding:NSASCIIStringEncoding]];
		++numLocObjsBeingSent;
	}
	[self->locationCache removeAllObjects];
	[postData appendData:[[NSString stringWithFormat:@"]"] dataUsingEncoding:NSASCIIStringEncoding]];

	// Write cached acclerometer data to the JSON string.
	[postData appendData:[[NSString stringWithFormat:@", \"accelerometer\": ["] dataUsingEncoding:NSASCIIStringEncoding]];
	size_t numAccelObjsBeingSent = 0;
	for (NSString* text in self->accelerometerCache)
	{
		if (numAccelObjsBeingSent > 0)
			[postData appendData:[[NSString stringWithFormat:@",\n"] dataUsingEncoding:NSASCIIStringEncoding]];
		[postData appendData:[text dataUsingEncoding:NSASCIIStringEncoding]];
		++numAccelObjsBeingSent;
	}
	[self->accelerometerCache removeAllObjects];
	[postData appendData:[[NSString stringWithFormat:@"]"] dataUsingEncoding:NSASCIIStringEncoding]];

	// Add the device ID to the JSON string.
	if (self->deviceId)
	{
		[postData appendData:[[NSString stringWithFormat:@",\n\"%s\":\"%@\"", KEY_NAME_DEVICE_ID, self->deviceId] dataUsingEncoding:NSASCIIStringEncoding]];
	}

	// Add the activity ID to the JSON string.
	if (activityId)
	{
		[postData appendData:[[NSString stringWithFormat:@",\n\"%s\":\"%@\"", KEY_NAME_ACTIVITY_ID, activityId] dataUsingEncoding:NSASCIIStringEncoding]];
	}

	// Add the activity type to the JSON string.
	char* activityType = GetCurrentActivityType();
	if (activityType)
	{
		[postData appendData:[[NSString stringWithFormat:@",\n\"%s\":\"%s\"", KEY_NAME_ACTIVITY_TYPE, activityType] dataUsingEncoding:NSASCIIStringEncoding]];
		free((void*)activityType);
	}

	// Add the user name to the JSON string.
	NSString* userName = [Preferences broadcastUserName];
	if (userName)
	{
		[postData appendData:[[NSString stringWithFormat:@",\n\"%s\":\"%@\"", ACTIVITY_ATTRIBUTE_USER_NAME, userName] dataUsingEncoding:NSASCIIStringEncoding]];
	}

	[postData appendData:[[NSString stringWithFormat:@"}\n"] dataUsingEncoding:NSASCIIStringEncoding]];

	if ((numLocObjsBeingSent > 0) || (numAccelObjsBeingSent > 0))
	{
		[self sendToServer:hostName withPath:REMOTE_API_UPDATE_STATUS_URL withData:postData withActivityId:activityId withActivityState:activityStopped];
	}

	self->lastCacheFlush = time(NULL);
}

- (void)broadcast
{
	// Flush at the user-specified interval. Default to 60 seconds if one was not specified.
	NSInteger rate = [Preferences broadcastRate];
	if (([self->locationCache count] > 0 || [self->accelerometerCache count] > 0) && (time(NULL) - self->lastCacheFlush > rate))
	{
		[self flushGlobalBroadcastCacheRest:FALSE withActivityId:[[NSString alloc] initWithFormat:@"%s", GetCurrentActivityId()]];
	}
}

/// @brief This method is called in response to an accelerometer updated notification.
- (void)accelerometerUpdated:(NSNotification*)notification
{
	if (![Preferences shouldBroadcastToServer])
	{
		return;
	}
	if (!IsActivityInProgress())
	{
		return;
	}
	if (!(IsLiftingActivity() || IsSwimmingActivity()))
	{
		return;
	}

	NSDictionary* accelerometerData = [notification object];

	NSError* error;
	NSData* jsonData = [NSJSONSerialization dataWithJSONObject:accelerometerData options:NSJSONWritingPrettyPrinted error:&error];
	NSString* text = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

	[self->accelerometerCache addObject:text];
	[self broadcast];
}

/// @brief This method is called in response to a location updated notification.
- (void)locationUpdated:(NSNotification*)notification
{
	if (![Preferences shouldBroadcastToServer])
	{
		return;
	}
	if (!IsActivityInProgress())
	{
		return;
	}
	if (!IsMovingActivity())
	{
		return;
	}

	NSDictionary* locationData = [notification object];
	NSMutableDictionary* broadcastData = [locationData mutableCopy];

	ActivityAttributeType attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_DISTANCE_TRAVELED);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_DISTANCE_TRAVELED];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_SPEED);
	if (attr.valid)
	{
		ConvertToBroadcastUnits(&attr);
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_AVG_SPEED];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_CURRENT_SPEED);
	if (attr.valid)
	{
		ConvertToBroadcastUnits(&attr);
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_CURRENT_SPEED];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_MOVING_SPEED);
	if (attr.valid)
	{
		ConvertToBroadcastUnits(&attr);
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_MOVING_SPEED];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_PACE);
	if (attr.valid)
	{
		ConvertToBroadcastUnits(&attr);
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_AVG_PACE];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_HEART_RATE);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_AVG_HEART_RATE];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_HEART_RATE);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_HEART_RATE];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_CADENCE);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_AVG_CADENCE];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_CADENCE);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_CADENCE];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_POWER);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_AVG_POWER];
	}

	attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_POWER);
	if (attr.valid)
	{
		NSNumber* value = [[NSNumber alloc] initWithDouble:attr.value.doubleVal];
		[broadcastData setObject:value forKey:@ACTIVITY_ATTRIBUTE_POWER];
	}

	NSError* error;
	NSData* jsonData = [NSJSONSerialization dataWithJSONObject:broadcastData options:NSJSONWritingPrettyPrinted error:&error];
	NSString* text = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

	[self->locationCache addObject:text];
	[self broadcast];
}

/// @brief This method is called in response to an activity started notification.
- (void)activityStarted:(NSNotification*)notification
{
	[self->locationCache removeAllObjects];
	[self->accelerometerCache removeAllObjects];
	self->lastCacheFlush = time(NULL);
}

/// @brief This method is called in response to an activity stopped notification.
- (void)activityStopped:(NSNotification*)notification
{
	@try
	{
		NSDictionary* activityData = [notification object];

		if (activityData)
		{
			// Make sure we've sent all the activity data.
			NSString* activityId = [activityData objectForKey:@KEY_NAME_ACTIVITY_ID];
			[self flushGlobalBroadcastCacheRest:TRUE withActivityId:activityId];
		}
	}
	@catch (...)
	{
	}
}

@end

#endif

File: IOS/Model/BroadcastManager.swift
//
//  BroadcastManager.swift
//  Created by Michael Simms on 9/29/22.
//

import Foundation

class BroadcastManager {
	static let shared = BroadcastManager()
	
	var locationCache: Array<String> = []      // Locations to be sent
	var accelerometerCache: Array<String> = [] // Accelerometer readings to be sent
	var lastCacheFlushTime: UInt64 = 0         // Unix time of the most recent cache flush
	var lastSendTime: UInt64 = 0               // Unix time of the most recent transmission
	var deviceId: String? = Preferences.uuid() // Unique identifier for the device doing the sending
	var dataBeingSent: String = ""             // Formatted data to be sent
	var errorSending: Bool = false             // Whether or not the last send was successful
	var currentActivityId: String = ""         // Cached for performance reasons
	var currentActivityType: String = ""       // Cached for performance reasons

	/// Singleton constructor
	private init() {
		NotificationCenter.default.addObserver(self, selector: #selector(self.accelerometerUpdated), name: Notification.Name(rawValue: NOTIFICATION_NAME_ACCELEROMETER), object: nil)
		NotificationCenter.default.addObserver(self, selector: #selector(self.locationUpdated), name: Notification.Name(rawValue: NOTIFICATION_NAME_LOCATION), object: nil)
		NotificationCenter.default.addObserver(self, selector: #selector(self.activityStarted), name: Notification.Name(rawValue: NOTIFICATION_NAME_ACTIVITY_STARTED), object: nil)
		NotificationCenter.default.addObserver(self, selector: #selector(self.activityStopped), name: Notification.Name(rawValue: NOTIFICATION_NAME_ACTIVITY_STOPPED), object: nil)
	}
	
	func getCurrentActivityId() -> String {
		let activityId = String(cString: UnsafeRawPointer(GetCurrentActivityId()).assumingMemoryBound(to: CChar.self))
		return activityId
	}
	
	func getCurrentActivityType() -> String {
		let activityType = String(cString: UnsafeRawPointer(GetCurrentActivityType()).assumingMemoryBound(to: CChar.self))
		return activityType
	}

	func displayMessage(text: String) {
	}

	func sendToServer(hostName: String, path: String, data: String, activityId: String, isStopped: Bool) {
		let protocolStr = Preferences.broadcastProtocol()
		let urlStr = String(format: "%@://%@/%@", protocolStr, hostName, path)
		let postLength = String(format: "%lu", data.count)
		
		var request = URLRequest(url: URL(string: urlStr)!)
		request.timeoutInterval = 30.0
		request.allowsExpensiveNetworkAccess = true
		request.httpMethod = "POST"
		request.setValue(postLength, forHTTPHeaderField:"Content-Length")
		request.setValue("application/json", forHTTPHeaderField:"Content-Type")
		request.httpBody = data.data(using: .utf8)
		
		let session = URLSession.shared
		let dataTask = session.dataTask(with: request) { data, response, error in
			if let httpResponse = response as? HTTPURLResponse {
				if httpResponse.statusCode == 200 {
					self.dataBeingSent = ""
					self.errorSending = false
					self.lastSendTime = UInt64(time(nil))
				}
				else {
					self.displayMessage(text: "Error sending data to the server.")
					self.errorSending = true
				}
			}
		}
		dataTask.resume()
	}

	func flushGlobalBroadcastCacheRest(isStopped: Bool) {
		// No host name set, just return.
		let hostName = Preferences.broadcastHostName()
		if hostName.count == 0 {
			NSLog("Broadcast host name not specified.")
			return
		}
		
		// Still waiting on last data to be sent.
		if self.dataBeingSent.count > 0 {
			if self.errorSending {
				self.sendToServer(hostName: hostName, path: REMOTE_API_UPDATE_STATUS_URL, data: self.dataBeingSent, activityId: self.currentActivityId, isStopped: isStopped)
				NSLog("Resending.")
			}
			else {
				NSLog("Waiting on previous data to be sent.")
			}
			self.lastCacheFlushTime = UInt64(time(nil))
			return
		}
		
		// Write cached location data to the JSON string.
		var post = "{\"locations\": ["
		var numLocObjsBeingSent = 0
		for text in self.locationCache {
			if numLocObjsBeingSent > 0 {
				post += ",\n"
			}
			post += text
			numLocObjsBeingSent += 1
		}
		self.locationCache.removeAll()
		post += "]"
		
		// Write cached accelerometer data to the JSON string.
		post += ", \"accelerometer\": ["
		var numAccelObjsBeingSent = 0
		for text in self.accelerometerCache {
			if numAccelObjsBeingSent > 0 {
				post += ",\n"
			}
			post += text
			numAccelObjsBeingSent += 1
		}
		self.accelerometerCache.removeAll()
		post += "]"
		
		// Add the device ID to the JSON string.
		if self.deviceId != nil && self.deviceId!.count > 0 {
			post += String(format: ",\n\"%@\":\"%@\"", KEY_NAME_DEVICE_ID, self.deviceId!)
		}
		
		// Add the activity ID to the JSON string.
		post += String(format: ",\n\"%@\":\"%@\"", KEY_NAME_ACTIVITY_ID, self.currentActivityId)
		
		// Add the activity type to the JSON string.
		post += String(format: ",\n\"%@\":\"%@\"", KEY_NAME_ACTIVITY_TYPE, self.currentActivityType)
		
		// Add the user name to the JSON string.
		let userName = Preferences.broadcastUserName()
		if userName.count > 0 {
			post += String(format: ",\n\"%@\":\"%@\"", ACTIVITY_ATTRIBUTE_USER_NAME, userName)
		}
		post += "}\n"
		
		if (numLocObjsBeingSent > 0) || (numAccelObjsBeingSent > 0) {
			self.sendToServer(hostName: hostName, path: REMOTE_API_UPDATE_STATUS_URL, data: post, activityId: self.currentActivityId, isStopped: isStopped)
		}
		
		self.lastCacheFlushTime = UInt64(time(nil))
	}
	
	func broadcast() {
		// Flush at the user-specified interval. Default to 60 seconds if one was not specified.
		let rate = Preferences.broadcastRate()
		
		// If we have data to send and it's been long enough since we last sent then flush the cache.
		if ((self.locationCache.count > 0 || self.accelerometerCache.count > 0) && (UInt64(time(nil)) - self.lastCacheFlushTime > rate)) {
			self.flushGlobalBroadcastCacheRest(isStopped: false)
		}
	}
	
	/// @brief This method is called in response to an accelerometer updated notification.
	@objc func accelerometerUpdated(notification: NSNotification) {
		if !Preferences.shouldBroadcastToServer() {
			return
		}
		if !IsActivityInProgress() {
			return
		}
		if !(IsLiftingActivity() || IsSwimmingActivity()) {
			return
		}

		if let accelerometerData = notification.object as? Dictionary<String, AnyObject> {

			do {
				let jsonData = try JSONSerialization.data(withJSONObject: accelerometerData, options: [])
				let text = String(data: jsonData, encoding: String.Encoding.ascii)!
				
				self.accelerometerCache.append(text)
				self.broadcast()
			}
			catch {
			}
		}
	}
	
	/// @brief This method is called in response to a location updated notification.
	@objc func locationUpdated(notification: NSNotification) {
		if !Preferences.shouldBroadcastToServer() {
			return
		}
		if !IsActivityInProgress() {
			return
		}
		if !IsMovingActivity() {
			return
		}
		
		if var locationData = notification.object as? Dictionary<String, AnyObject> {

			do {
				var attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_DISTANCE_TRAVELED)
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_DISTANCE_TRAVELED] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_SPEED);
				if attr.valid {
					ConvertToBroadcastUnits(&attr)
					locationData[ACTIVITY_ATTRIBUTE_AVG_SPEED] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_CURRENT_SPEED);
				if attr.valid {
					ConvertToBroadcastUnits(&attr);
					locationData[ACTIVITY_ATTRIBUTE_CURRENT_SPEED] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_MOVING_SPEED);
				if attr.valid {
					ConvertToBroadcastUnits(&attr);
					locationData[ACTIVITY_ATTRIBUTE_MOVING_SPEED] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_PACE);
				if attr.valid {
					ConvertToBroadcastUnits(&attr);
					locationData[ACTIVITY_ATTRIBUTE_AVG_PACE] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_HEART_RATE);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_AVG_HEART_RATE] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_HEART_RATE);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_HEART_RATE] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_CADENCE);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_AVG_CADENCE] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_CADENCE);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_CADENCE] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_AVG_POWER);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_AVG_POWER] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_POWER);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_POWER] = attr.value.doubleVal as AnyObject?
				}
				
				attr = QueryLiveActivityAttribute(ACTIVITY_ATTRIBUTE_THREAT_COUNT);
				if attr.valid {
					locationData[ACTIVITY_ATTRIBUTE_THREAT_COUNT] = attr.value.doubleVal as AnyObject?
				}
				
				let jsonData = try JSONSerialization.data(withJSONObject: locationData, options: [])
				let text = String(data: jsonData, encoding: String.Encoding.ascii)!
				
				self.locationCache.append(text)
				self.broadcast()
			}
			catch {
			}
		}
	}

	/// @brief This method is called in response to an activity started notification.
	@objc func activityStarted(notification: NSNotification) {
		self.locationCache.removeAll()
		self.accelerometerCache.removeAll()
		self.lastCacheFlushTime = UInt64(time(nil))
		self.lastSendTime = 0
		self.deviceId = Preferences.uuid()
		self.currentActivityId = self.getCurrentActivityId()
		self.currentActivityType = self.getCurrentActivityType()
	}

	/// @brief This method is called in response to an activity stopped notification.
	@objc func activityStopped(notification: NSNotification) {
		self.flushGlobalBroadcastCacheRest(isStopped: true)
	}
}
